compiler/hash-typecheck/src/exhaustiveness/lower.rs

```diff
@@ -135,123 +54,319 @@ impl<'tc> LowerPatOps<'tc> {
-    pub fn lower_constructor(&self, pat: primitives::Pat, span: Span) -> Pat {
-        let ConstructorPat { subject, args } = match pat {
-            primitives::Pat::Constructor(constructor) => constructor,
-            _ => unreachable!(),
-        };
+        // Now we need to put them in the store...
+        let fields = Fields::from_iter(
+            fields.into_iter().map(|field| self.deconstructed_pat_store().create(field)),
+        );
 
-        // Transform the arguments into fields, since it doesn't matter
-        // whether this will become a variant or a leaf.
-        let pats = self.lower_pat_fields(args);
+        DeconstructedPat::new(ctor, fields, ty, Some(pat_id))
+    }
 
-        // We need to determine if this is a enumeration or a struct, if it is a
-        // struct, we can easily conclude that this lowered pattern is a `Leaf`,
-        // otherwise it becomes a variant kind
+    // Convert a [DeconstructedPat] into a [Pat].
+    pub(crate) fn construct_pat(&self, pat: DeconstructedPatId) -> PatId {
         let reader = self.reader();
-        let term = reader.get_term(subject);
+        let pat = reader.get_deconstructed_pat(pat);
+
+        // Short-circuit, if the pattern already has an associated `PatId`...
+        if pat.id.is_some() {
+            return pat.id.unwrap();
+        }
+
+        let ctor = reader.get_ctor(pat.ctor);
 
-        let kind = match term {
-            primitives::Term::Level1(Level1Term::NominalDef(id)) => {
-                let nominal_def = reader.get_nominal_def(*id);
+        // Build the pattern based from the constructor and the fields...
+        let pat = match ctor {
+            DeconstructedCtor::Single | DeconstructedCtor::Variant(_) => {
+                let reader = self.reader();
+
+                match reader.get_term(pat.ty) {
+                    Term::Level1(Level1Term::Tuple(_)) => {
+                        let _children = pat
+                            .fields
+                            .iter_patterns()
+                            .map(|p| self.construct_pat(*p))
+                            .collect_vec();
+
+                        // let args = self.builder().create_pat_args(children);
+                        // Pat::Tuple(args)
+
+                        todo!()
+                    }
+                    Term::Level1(Level1Term::NominalDef(_)) => {
+                        let _children = pat.fields.iter_patterns().map(|p| self.construct_pat(*p));
 
-                match nominal_def {
-                    NominalDef::Struct(_) => PatKind::Leaf { pats },
-                    // @@EnumToUnion: get the variant index here
-                    NominalDef::Enum(_) => unreachable!(),
+                        // let args = self.builder().create_pat_args(children);
+                        // Pat::Constructor(ConstructorPat { subject: pat.ty, args })
+                        todo!()
```

Comment:
todo left

Reply:
The todo is dependant on the `builder` having immutable access

Reply:
Maybe write a `@@Todo: immutable builder`?